scripts/west_commands/bindesc.py

```diff
@@ -0,0 +1,314 @@
+# Copyright (c) 2023 Yonatan Schachter
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from textwrap import dedent
+import struct
+
+from west.commands import WestCommand
+from west import log
+
+
+try:
+    from elftools.elf.elffile import ELFFile
+    from intelhex import IntelHex
+    MISSING_REQUIREMENTS = False
+except ImportError:
+    MISSING_REQUIREMENTS = True
+
+
+# Based on scripts/build/uf2conv.py
+def convert_from_uf2(buf):
+    UF2_MAGIC_START0 = 0x0A324655 # First magic number ('UF2\n')
+    UF2_MAGIC_START1 = 0x9E5D5157 # Second magic number
+    numblocks = len(buf) // 512
+    curraddr = None
+    outp = []
+    for blockno in range(numblocks):
+        ptr = blockno * 512
+        block = buf[ptr:ptr + 512]
+        hd = struct.unpack(b'<IIIIIIII', block[0:32])
+        if hd[0] != UF2_MAGIC_START0 or hd[1] != UF2_MAGIC_START1:
+            log.inf('Skipping block at ' + ptr + '; bad magic')
+            continue
+        if hd[2] & 1:
+            # NO-flash flag set; skip block
+            continue
+        datalen = hd[4]
+        if datalen > 476:
+            log.die(f'Invalid UF2 data size at {ptr}')
+        newaddr = hd[3]
+        if curraddr is None:
+            curraddr = newaddr
+        padding = newaddr - curraddr
+        if padding < 0:
+            log.die(f'Block out of order at {ptr}')
+        if padding > 10*1024*1024:
+            log.die(f'More than 10M of padding needed at {ptr}')
+        if padding % 4 != 0:
+            log.die(f'Non-word padding size at {ptr}')
+        while padding > 0:
+            padding -= 4
+            outp += b'\x00\x00\x00\x00'
+        outp.append(block[32 : 32 + datalen])
+        curraddr = newaddr + datalen
+    return b''.join(outp)
+
+
+class Bindesc(WestCommand):
+    EXTENSIONS = ['bin', 'hex', 'elf', 'uf2']
+
+    # Corresponds to the definitions in include/zephyr/bindesc.h.
+    # Do not change without syncing the definitions in both files!
+    TYPE_UINT = 0
+    TYPE_STR = 1
+    TYPE_BYTES = 2
+    MAGIC = 0x2c31b9b9
+    DESCRIPTORS_END = 0xffff
+
+    def __init__(self):
+        self.TAG_TO_NAME = {
+            # Corresponds to the definitions in include/zephyr/bindesc.h.
+            # Do not change without syncing the definitions in both files!
+            self.bindesc_gen_tag(self.TYPE_STR, 0x800): 'APP_VERSION_STRING',
+            self.bindesc_gen_tag(self.TYPE_UINT, 0x801): 'APP_VERSION_MAJOR',
+            self.bindesc_gen_tag(self.TYPE_UINT, 0x802): 'APP_VERSION_MINOR',
+            self.bindesc_gen_tag(self.TYPE_UINT, 0x803): 'APP_VERSION_REVISION',
```

Comment:
And here, I don't see this field, I see BINDESC_ID_APP_VERSION_PATCHLEVEL and BINDESC_ID_APP_VERSION_NUMBER